pkg/client/testing/core/fixture.go

```diff
@@ -49,153 +58,158 @@ type ObjectScheme interface {
+			return true, obj, err
 
-		case CreateAction:
-			accessor, err := meta.Accessor(castAction.GetObject())
+		case UpdateActionImpl:
+			objMeta, err := meta.Accessor(typedAction.GetObject())
+			if err != nil {
+				return true, nil, err
+			}
+			err = registry.Update(typedAction.GetObject())
 			if err != nil {
 				return true, nil, err
 			}
-			resource, err := o.Kind(kind, accessor.GetName())
-			return true, resource, err
+			obj, err := registry.Get(kind, ns, objMeta.GetName())
+			return true, obj, err
 
-		case UpdateAction:
-			accessor, err := meta.Accessor(castAction.GetObject())
+		case DeleteActionImpl:
+			err := registry.Delete(kind, ns, typedAction.GetName())
 			if err != nil {
 				return true, nil, err
 			}
-			resource, err := o.Kind(kind, accessor.GetName())
-			return true, resource, err
+			return true, nil, nil
 
 		default:
 			return false, nil, fmt.Errorf("no reaction implemented for %s", action)
 		}
 	}
 }
 
-// AddObjectsFromPath loads the JSON or YAML file containing Kubernetes API resources
-// and adds them to the provided ObjectRetriever.
-func AddObjectsFromPath(path string, o ObjectRetriever, decoder runtime.Decoder) error {
-	data, err := ioutil.ReadFile(path)
-	if err != nil {
-		return err
-	}
-	data, err = yaml.ToJSON(data)
-	if err != nil {
-		return err
-	}
-	obj, err := runtime.Decode(decoder, data)
-	if err != nil {
-		return err
-	}
-	if err := o.Add(obj); err != nil {
-		return err
-	}
-	return nil
-}
-
-type objects struct {
-	types   map[string][]runtime.Object
-	last    map[string]int
+type registry struct {
 	scheme  ObjectScheme
 	decoder runtime.Decoder
+	mu      sync.RWMutex
+	objects map[string][]runtime.Object
 }
 
-var _ ObjectRetriever = &objects{}
-
-// NewObjects implements the ObjectRetriever interface by introspecting the
-// objects provided to Add() and returning them when the Kind method is invoked.
-// If an api.List object is provided to Add(), each child item is added. If an
-// object is added that is itself a list (PodList, ServiceList) then that is added
-// to the "PodList" kind. If no PodList is added, the retriever will take any loaded
-// Pods and return them in a list. If an api.Status is added, and the Details.Kind field
-// is set, that status will be returned instead (as an error if Status != Success, or
-// as a runtime.Object if Status == Success).  If multiple PodLists are provided, they
-// will be returned in order by the Kind call, and the last PodList will be reused for
-// subsequent calls.
-func NewObjects(scheme ObjectScheme, decoder runtime.Decoder) ObjectRetriever {
-	return objects{
-		types:   make(map[string][]runtime.Object),
-		last:    make(map[string]int),
+var _ Registry = &registry{}
+
+// NewObjects returns a Registry that can be used to keep track of
+// objects for the fake clientset. Mostly useful for unit tests.
+func NewObjects(scheme ObjectScheme, decoder runtime.Decoder) Registry {
```

Comment:
I would love to, but that will require regenerating a fake clientset code. Do you think it's worth it?